net: tracing: Add socket tracing support

jukkar opened the pull request:
If network socket tracing is enabled, then the system will track various socket API calls for usage.

jukkar committed:
net: tracing: Add socket tracing support

If network socket tracing is enabled, then the system will track
various socket API calls for usage.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

nashif commented:
@jukkar  nice, while playing with this, discovered some missing hooks and added them here: #73674, not sure if you hit the same issue.

jukkar commented:
> not sure if you hit the same issue.

No, I did not see any issue with those missing hooks. I will take your commit to this PR in order to avoid conflicts later.

jukkar committed:
tests: net: socket: Avoid macro error with zsock_sendto

The preprocessor gives this error when calling zsock_sendto
in user mode when there is a mismatch with amount of parameters
to the sendto wrapper call.

  tests/net/socket/udp/src/main.c: In function ‘comm_sendto_recvfrom’:
  tests/net/socket/udp/src/main.c:125:46: error: macro "zsock_sendto" \
                                 requires 6 arguments, but only 5 given
  125 |                             0, &addr, addrlen);
      |                                              ^
  In file included from include/zephyr/net/socket.h:1381,
                 from tests/net/socket/udp/src/main.c:15:
  include/generated/zephyr/syscalls/socket.h:266: note: \
                      macro "zsock_sendto" defined here
  266 | #define zsock_sendto(sock, buf, len, flags, dest_addr, addrlen)

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

nashif commented:
> No, I did not see any issue with those missing hooks. I will take your commit to this PR in order to avoid conflicts later.

I got this with tests/net/socket/tcp/ and enabling tracing...

cfriedt reviewed (dismissed):
🪁

jukkar commented:
> I got this with tests/net/socket/tcp/ and enabling tracing...

Thanks, that issue is now fixed

rlubos commented on include/zephyr/tracing/tracing.h:
```diff
@@ -2026,6 +2026,304 @@
+#define sys_port_trace_socket_accept_enter(socket)
+
+/**
+ * @brief Trace network socket listen attempt
```
s/listen/accept

rlubos commented on include/zephyr/tracing/tracing.h:
```diff
@@ -2026,6 +2026,304 @@
+#define sys_port_trace_socket_recvfrom_enter(socket, max_len, flags, addr, addrlen)
+
+/**
+ * @brief Trace network socket sendmsg attempt
```
s/sendmsg/recvfrom

rlubos commented on include/zephyr/tracing/tracing.h:
```diff
@@ -2026,6 +2026,304 @@
+#define sys_port_trace_socket_setsockopt_enter(socket, level, optname, optval, optlen)
+
+/**
+ * @brief Trace network socket getsockopt attempt
```
s/getsockopt/setsockopt

rlubos reviewed (dismissed):
LGTM, just a few nits in docs

jukkar commented:
I ran the tests/net twister suite manually with tracing enabled, and did not spot any issues any more.

jukkar commented:
Twister fails with unrelated issues in the tree, upstream seems to be broken again
```
west build -p -b native_sim tests/subsys/sensing -T subsys.sensing -d ../build/sensing -t run
...
[113/115] Building native simulator runner, and linking final executable
FAILED: zephyr/CMakeFiles/native_runner_executable zephyr/zephyr.exe .../zephyrproject/build/sensing/zephyr/CMakeFiles/native_runner_executable .../zephyrproject/build/sensing/zephyr/zephyr.exe 
cd .../zephyrproject/build/sensing/zephyr && /usr/bin/make -f .../zephyrproject/zephyr/scripts/native_simulator/Makefile all --warn-undefined-variables -r NSI_CONFIG_FILE=.../zephyrproject/build/sensing/zephyr/NSI/nsi_config
/usr/bin/ld: .../zephyrproject/build/sensing/zephyr/NSI/.../zephyrproject/build/sensing/zephyr/zephyr.elf.loc_cpusw.o:.../zephyrproject/zephyr/drivers/sensor/bosch/bmi160/bmi160.c:1233: undefined reference to `__device_dts_ord_35'
/usr/bin/ld: .../zephyrproject/build/sensing/zephyr/NSI/.../zephyrproject/build/sensing/zephyr/zephyr.elf.loc_cpusw.o:.../zephyrproject/zephyr/drivers/sensor/bosch/bmi160/bmi160.c:1233: undefined reference to `__device_dts_ord_32'
collect2: error: ld returned 1 exit status
```

jukkar commented:
ping @nashif, are you ok with the changes?
